### DISCUSSION TITLE
Can't start mariadb on Raspberry Pi 3 (32-bit armv7)

### DISCUSSION THREAD

**Discussion by nigelorr** [photoprism/photoprism#1889] [Q&A] [answered]
I'm trying out photoprism, aiming to use it to manage an existing collection of 40,000 images. I've made a test installation on a Raspberry Pi 3 to see if it will work for my needs with a small subset of the photos, but it's not working (expecting to run on a Pi 4 or something faster for the actual setup).

I've followed the Pi's [system requirements](https://docs.photoprism.app/getting-started/raspberry-pi/#system-requirements) and the [armv7 specific instructions](https://docs.photoprism.app/getting-started/docker-compose/), and used the [docker-compose configuration for armv7](https://dl.photoprism.app/docker/armv7/docker-compose.yml). I've added 4GB swap to the Pi.

The first problem I had with the docker-compose configuration was that mariadb tries to run as root, so exits. I modified the command in the mariadb section to add `--user=mysql` so it now tries to run. That seems odd, as it would affect anyone using that file and I guess others will have used it before me?

Now, it fails with ` Fatal error: Can't open and lock privilege tables: Table 'mysql.db' doesn't exist`. If I change the command to replace `command: mysqld --inodb...` with a dummy `command: sh -c "sleep 99999999"`, and connect to the image using ` docker-compose exec mariadb bash`, I can have a look around the running container.

If I run `mariadb -u root -p` from there, I can log in to the database and see that mysql.db is populated. However, as I have commented out the command that starts mysqld I didn't expect it to be running, does that give a clue to the problem, is the normal `command` definition in the yml file trying to start mariadb mysqld which is already running in the docker?

I've gone through the [troubleshooting guide](https://docs.photoprism.app/getting-started/troubleshooting/#fatal-server-errors) and everything seems OK (permissions to /config are open to all and owned by mysql etc). Can anyone help me to get this working?

Logs show a repeated loop of:
> mariadb_1     | [s6-init] making user provided files available at /var/run/s6/etc...exited 0.
> mariadb_1     | [s6-init] ensuring user provided files have correct perms...exited 0.
> mariadb_1     | [fix-attrs.d] applying ownership & permissions fixes...
> mariadb_1     | [fix-attrs.d] done.
> mariadb_1     | [cont-init.d] executing container initialization scripts...
> mariadb_1     | [cont-init.d] 01-envfile: executing...
> mariadb_1     | [cont-init.d] 01-envfile: exited 0.
> mariadb_1     | [cont-init.d] 10-adduser: executing...
> mariadb_1     | usermod: no changes
> mariadb_1     |
> mariadb_1     | -------------------------------------
> mariadb_1     |           _         ()
> mariadb_1     |          | |  ___   _    __
> mariadb_1     |          | | / __| | |  /  \
> mariadb_1     |          | | \__ \ | | | () |
> mariadb_1     |          |_| |___/ |_|  \__/
> mariadb_1     |
> mariadb_1     |
> mariadb_1     | Brought to you by linuxserver.io
> mariadb_1     | -------------------------------------
> mariadb_1     |
> mariadb_1     | To support LSIO projects visit:
> mariadb_1     | https://www.linuxserver.io/donate/
> mariadb_1     | -------------------------------------
> mariadb_1     | GID/UID
> mariadb_1     | -------------------------------------
> mariadb_1     |
> mariadb_1     | User uid:    101
> mariadb_1     | User gid:    100
> mariadb_1     | -------------------------------------
> mariadb_1     |
> mariadb_1     | [cont-init.d] 10-adduser: exited 0.
> mariadb_1     | [cont-init.d] 30-config: executing...
> mariadb_1     | [cont-init.d] 30-config: exited 0.
> mariadb_1     | [cont-init.d] 40-initialise-db: executing...
> mariadb_1     | [cont-init.d] 40-initialise-db: exited 0.
> mariadb_1     | [cont-init.d] 90-custom-folders: executing...
> mariadb_1     | [cont-init.d] 90-custom-folders: exited 0.
> mariadb_1     | [cont-init.d] 99-custom-files: executing...
> mariadb_1     | [custom-init] no custom files found exiting...
> mariadb_1     | [cont-init.d] 99-custom-files: exited 0.
> mariadb_1     | [cont-init.d] done.
> mariadb_1     | [services.d] starting services
> mariadb_1     | [services.d] done.
> mariadb_1     | 2022-01-05 17:33:36 0 [Note] mysqld (mysqld 10.5.13-MariaDB-log) starting as process 305 ...
> mariadb_1     | 2022-01-05 17:33:36 0 [Note] InnoDB: Uses event mutexes
> mariadb_1     | 2022-01-05 17:33:36 0 [Note] InnoDB: Compressed tables use zlib 1.2.11
> mariadb_1     | 2022-01-05 17:33:36 0 [Note] InnoDB: Number of pools: 1
> mariadb_1     | 2022-01-05 17:33:36 0 [Note] InnoDB: Using generic crc32 instructions
> mariadb_1     | 2022-01-05 17:33:36 0 [Note] mysqld: O_TMPFILE is not supported on /var/tmp (disabling future attempts)
> mariadb_1     | 220105 17:33:36 mysqld_safe Logging to '/config/databases/5ba75db37aab.err'.
> mariadb_1     | 2022-01-05 17:33:36 0 [Note] InnoDB: Using Linux native AIO
> mariadb_1     | 2022-01-05 17:33:36 0 [Note] InnoDB: Initializing buffer pool, total size = 134217728, chunk size = 134217728
> mariadb_1     | 2022-01-05 17:33:36 0 [Note] InnoDB: Completed initialization of buffer pool
> mariadb_1     | 220105 17:33:36 mysqld_safe Starting mariadbd daemon with databases from /config/databases
> mariadb_1     | 2022-01-05 17:33:36 0 [Note] InnoDB: 128 rollback segments are active.
> mariadb_1     | 2022-01-05 17:33:36 0 [Note] InnoDB: Creating shared tablespace for temporary tables
> mariadb_1     | 2022-01-05 17:33:36 0 [Note] InnoDB: Setting file './ibtmp1' size to 12 MB. Physically writing the file full; Please wait ...
> mariadb_1     | 2022-01-05 17:33:36 0 [Note] InnoDB: File './ibtmp1' size is now 12 MB.
> mariadb_1     | 2022-01-05 17:33:36 0 [Note] InnoDB: 10.5.13 started; log sequence number 35040; transaction id 8
> mariadb_1     | 2022-01-05 17:33:36 0 [Note] InnoDB: Loading buffer pool(s) from /var/lib/mysql/ib_buffer_pool
> mariadb_1     | 2022-01-05 17:33:36 0 [Note] Plugin 'FEEDBACK' is disabled.
> mariadb_1     | 2022-01-05 17:33:36 0 [ERROR] Could not open mysql.plugin table: "Table 'mysql.plugin' doesn't exist". Some plugins may be not loaded
> mariadb_1     | 2022-01-05 17:33:36 1 [Warning] Failed to load slave replication state from table mysql.gtid_slave_pos: 1017: Can't find file: './mysql/' (errno: 2 "No such file or directory")
> mariadb_1     | 2022-01-05 17:33:36 0 [Note] InnoDB: Buffer pool(s) load completed at 220105 17:33:36
> mariadb_1     | 2022-01-05 17:33:37 0 [ERROR] Can't open and lock privilege tables: Table 'mysql.servers' doesn't exist
> mariadb_1     | 2022-01-05 17:33:37 0 [Note] Server socket created on IP: '::'.
> mariadb_1     | 2022-01-05 17:33:37 0 [ERROR] Fatal error: Can't open and lock privilege tables: Table 'mysql.db' doesn't exist
> mariadb_1     | 2022-01-05 17:33:37 0 [ERROR] Aborting
> mariadb_1     | [cmd] mysqld exited 1
> mariadb_1     | Caught SIGTERM signal!
> mariadb_1     | cat: /var/run/mysqld/mysqld.pid: No such file or directory
> mariadb_1     | [cont-finish.d] executing container finish scripts...
> mariadb_1     | [cont-finish.d] done.
> mariadb_1     | [s6-finish] waiting for services.
> mariadb_1     | [s6-finish] sending all processes the TERM signal.
> mariadb_1     | [s6-finish] sending all processes the KILL signal and exiting.

**Comment by lastzero** [comment]
Can you contact the maintainers or community members of the mariadb docker image? 

https://hub.docker.com/r/linuxserver/mariadb

It's not from us and we already have a lot of work, so no time to provide support for linuxserver 32-bit ARMv7 images. We also don't have a Pi 3 to test.

Since it's the first time this is being reported, there are probably users out there for whom it works.

**Comment by nigelorr** [comment, marked as answer]
Thanks for the pointer, I thought it must be a photoprism-specific problem. I reached out to the linuxserver folks on discord and got great support from them too.

The problem is in the armv7 docker-compose.yml file. It shouldn't have the line `command = mysql...` in the mariadb section. The linuxserver image you call out for armv7 starts mariadb automatically so that was trying to start a duplicate process.

Like you, I presumed others would have had it working so spent a day double-checking and triple-checking the instructions, so hopefully this will help the next person.

Hope you can fix the yml accordingly, or at least that this will be useful to anyone else running on a Pi 3 (I guess I'm the first as that configuration couldn't have worked?). I found that the system 'bogged down' and was unusable on the Pi 3, so have moved to a Pi 4 running 64-bit debian bullseye for the test instead now, that installed very easily and is busy importing photos.

Thanks for the quick reply

**Reply by lastzero** [reply]
Thanks for the hint! Our config example runs mariadb using the same docker-compose file, so you can simply use the service name to connect. Also there should be another way to configure mariadb then. Wondering why they do everything different than the standard Docker image? :dancers:

**Reply by lastzero** [reply]
Updated: https://dl.photoprism.app/docker/armv7/docker-compose.yml
